source-postgres/fdbutils.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/estuary/flow/go/protocols/fdb/tuple"
+	"github.com/jackc/pgtype"
+)
+
+const numericKey = "N"
+
+// EncodePgNumericKeyFDB encodes a pgtype.Numeric value to an FDB tuple.
+func EncodePgNumericKeyFDB(key pgtype.Numeric) (tuple.Tuple, error) {
+	if encodedKey, err := encodePgNumeric(key); err != nil {
+		return nil, fmt.Errorf("encode pgtype.Numeric: %w", err)
+	} else if encodedBinary, err := key.EncodeBinary(nil, nil); err != nil {
+		return nil, fmt.Errorf("encode binary pgtype Numeric: %w", err)
+	} else {
+		return tuple.Tuple{numericKey, encodedKey, encodedBinary}, nil
+	}
+}
+
+// DecodeFDBTuple decodes the input TupleElement to its original value.
+// Return nil if the input cannot be decoded.
+func DecodeFDBTuple(t tuple.TupleElement) interface{} {
+	switch t := t.(type) {
+	case tuple.Tuple:
+		if len(t) == 3 {
+			if key, ok := t[0].(string); ok && key == numericKey {
+				if b, ok := t[2].([]byte); ok {
+					var n pgtype.Numeric
+					if err := n.DecodeBinary(nil, b); err == nil {
+						return n
+					}
+				}
+			}
+		}
+	}
+	return nil
```

Comment:
If the input cannot be decoded this should probably be a fatal error. These tuple values should only ever come from our own encoding logic, so a failure here means that either a logic bug or some corruption of the actual encoded key has taken place.

If we silently replace that with a `nil` and the user has specified their own scan key which includes a nullable column, this `nil` will be used as the actual resume column value for backfills and could result in us missing or duplicating certain rows in the capture.

Reply:
I see. Here needs some explanations,
`DecodeFDBTuple` is called by `DecodeKeyFDB`, which is the external API of the database.
- The logic of `DecodeFDBTuple` is that - if it can decode the tuple, it will return the interpretation results. And if it cannot, it will return nil, indicating that `I cannot interpret it`.

- The logic of `DecodeKeyFDB` is that - if `DecodeFDBTuple` (or any other API in the future) can interpret the tuple, it will return the interpretation results, and if not, it will return the input directly without any interpretations.

So the external  API calls to `DecodeFDBTuple`  never gets a `nil` if the input is not `nil`. And this also explains why `DecodeFDBTuple` does not return any error, b/c if anything results an error, `DecodeFDBTuple` will simply return the input.

(I think the logic might be convoluted, which result in the mis-understandings. I will think about and improve it. There could be a restriction on if the input is a `tuple.Tuple`, it must be interpreted to sth.)